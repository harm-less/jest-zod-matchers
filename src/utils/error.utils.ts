import {SafeParseError, SafeParseReturnType} from 'zod';
import {fromZodError} from 'zod-validation-error';

export function isSafeParseError(
	receivedSafeParseReturnType: SafeParseReturnType<unknown, unknown>
): receivedSafeParseReturnType is SafeParseError<unknown> {
	return receivedSafeParseReturnType.success === false;
}

export function handleSafeParseError(
	context: jest.MatcherContext,
	receivedResponse: SafeParseError<unknown>,
	{
		matcherHint
	}: {
		matcherHint: string;
	}
): jest.CustomMatcherResult {
	const receivedErrorMessage = getCleanZodError(receivedResponse).message;

	return {
		message: () =>
			matcherHint +
			'\n\n' +
			'Expected Zod schema input to be valid but a Zod error occured instead.' +
			'\n\n' +
			context.utils.BOLD_WEIGHT(`Zod error: ${receivedErrorMessage}`),
		pass: false
	};
}

export function getCleanZodError(receivedError: SafeParseError<unknown>) {
	return fromZodError(receivedError.error, {
		maxIssuesInMessage: 1,
		prefix: '',
		prefixSeparator: ''
	});
}
