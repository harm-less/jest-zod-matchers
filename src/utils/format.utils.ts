export function formatExpectedResult(obj: Record<string, unknown>): Record<string, unknown> {
	for (const propertyName in obj) {
		if (obj.hasOwnProperty(propertyName)) {
			const propertyValue = obj[propertyName];
			if (propertyValue instanceof Date) {
				obj[propertyName] = propertyValue.toISOString();
			} else if (typeof propertyValue === 'object' && propertyValue) {
				formatExpectedResult(propertyValue as Record<string, unknown>);
			}
		}
	}
	return obj;
}
