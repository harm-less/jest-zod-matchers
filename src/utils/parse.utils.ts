import {ParseParams, ZodRawShape, z} from 'zod';

export function parseObjectSchema(
	objectSchema: z.ZodObject<ZodRawShape>,
	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	value: any,
	params?: Partial<ParseParams>
) {
	return objectSchema.safeParse(value, params);
}
