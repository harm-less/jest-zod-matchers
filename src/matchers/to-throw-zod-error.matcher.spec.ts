import {z} from 'zod';
import {parseObjectSchema} from '../utils/parse.utils';
import {toThrowZodError} from './to-throw-zod-error.matcher';

describe('"zodParseToThrowError" matcher', () => {
	let FooSchema: z.ZodObject<{
		foo: z.ZodString;
	}>;

	beforeEach(() => {
		expect.extend({toThrowZodError});

		FooSchema = z.object({
			foo: z.string()
		});
	});

	it('should have the expected Zod error due to an incorrect schema', () => {
		expect(parseObjectSchema(FooSchema, {foo: 10})).toThrowZodError(
			'Expected string, received number at \"foo\"'
		);
	});

	it('should not have the expected Zod error due to an incorrect schema', () => {
		expect(parseObjectSchema(FooSchema, {foo: 10})).not.toThrowZodError('Other error');
	});

	describe('errors', () => {
		it('should throw an error as the input is valid', () => {
			expect(() =>
				expect(parseObjectSchema(FooSchema, {foo: 'bar'})).toThrowZodError(
					'This will not be triggered'
				)
			).toThrowErrorMatchingSnapshot();
		});

		it('should throw an error as the Zod error should not be the same as the expected error', () => {
			expect(() =>
				expect(parseObjectSchema(FooSchema, {foo: 10})).not.toThrowZodError(
					'Expected string, received number at \"foo\"'
				)
			).toThrowErrorMatchingSnapshot();
		});
	});
});
