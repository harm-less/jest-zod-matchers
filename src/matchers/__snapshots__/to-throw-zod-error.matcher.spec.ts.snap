// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`"zodParseToThrowError" matcher errors should throw an error as the Zod error should not be the same as the expected error 1`] = `
"[2mexpect([22m[31mzodSafeParseReturnType[39m[2m).[22mnot[2m.[22mtoThrowZodError[2m([22m[32mexpectedErrorMessage[39m[2m) // Zod schema validation error message equality[22m

Expected a Zod error to have been thrown, but it should not equal the expected error

Expected: [32m"Expected string, received number at \\"foo\\""[39m
Received: serializes to the same string"
`;

exports[`"zodParseToThrowError" matcher errors should throw an error as the input is valid 1`] = `
"[2mexpect([22m[31mzodSafeParseReturnType[39m[2m).[22mtoThrowZodError[2m([22m[32mexpectedErrorMessage[39m[2m) // Zod schema validation error message equality[22m

Expected an error to have been thrown with the following message: 
[32m"This will not be triggered"[39m 

But no error was throw."
`;
